File: test-api-pkg/ts/test-docs-resources.ts
/**
 * Mirrors website/public/sdk/resources.md against @imbrace/sdk@1.0.4 (npm)
 * — API-key auth.
 *
 * Sections:
 *   §1 Contacts          (list / get / update / getComments / getFiles / getActivities)
 *   §2 Conversations     (search / getOutstanding / assignTeamMember / updateStatus)
 *   §3 Messaging         (channel.list / messages.send / messages.list)
 *   §4 Campaign + touchpoints (CRUD)
 *   §5 Message suggestion (getSuggestions)
 *   §6 Predict
 *
 * Fixtures are auto-resolved from list endpoints when possible. Destructive
 * mutations against prod data (contact.update, conversation.assign/updateStatus,
 * messages.send) are skipped. The campaign lifecycle is exercised on a
 * throwaway campaign that is deleted at the end.
 */
import "dotenv/config"
import { ImbraceClient } from "@imbrace/sdk"

const apiKey         = process.env.IMBRACE_API_KEY
const organizationId = process.env.IMBRACE_ORGANIZATION_ID
const baseUrl        = process.env.IMBRACE_GATEWAY_URL ?? "https://app-gatewayv2.imbrace.co"

if (!apiKey || !organizationId) {
  console.error("Missing IMBRACE_API_KEY or IMBRACE_ORGANIZATION_ID")
  process.exit(1)
}

// Generous timeout because contacts/channel/campaign/predict routes go to
// services (channel-service, campaign-service) flagged as unreachable on
// app-gatewayv2 in FIX_PLAN_v1.0.6.md §A.1. Keep at 30s so we differentiate
// "service truly down" from "transient slow".
const client = new ImbraceClient({ apiKey, organizationId, baseUrl, timeout: 30_000 })

let pass = 0, fail = 0, skip = 0
const fails: string[] = []
const docGaps: string[] = []

async function step(label: string, fn: () => Promise<unknown>, expectFail = false) {
  process.stdout.write(`  • ${label} ... `)
  try {
    const t0 = Date.now()
    const r = await fn()
    const dt = Date.now() - t0
    const summary = JSON.stringify(r ?? {}).slice(0, 90)
    if (expectFail) { console.log(`unexpected pass [${dt}ms]: ${summary}`); fail++; fails.push(`${label} → unexpected pass`) }
    else { console.log(`✓ [${dt}ms] ${summary}`); pass++ }
  } catch (err: any) {
    const code = err?.statusCode ?? err?.message ?? "ERR"
    if (expectFail) { console.log(`✓ (expected fail [${code}])`); pass++ }
    else { console.log(`✗ [${code}]`); fail++; fails.push(`${label} → ${code}`) }
  }
}

function skipped(label: string, reason: string) { console.log(`  - ${label}  ⏭ ${reason}`); skip++ }
function section(title: string) { console.log(`\n══ ${title} ══`) }
function note(msg: string) { console.log(`  ℹ ${msg}`); docGaps.push(msg) }

console.log(`\n━━━ DOCS: resources.md — auth: API KEY (npm @imbrace/sdk@1.0.4) ━━━`)
console.log(`gateway=${baseUrl}\norg=${organizationId}\n`)

const ts = Date.now()
const state: { contactId: string | null; convId: string | null; buId: string | null; campaignId: string | null; touchpointId: string | null } =
  { contactId: null, convId: null, buId: null, campaignId: null, touchpointId: null }

// ── §1. Contacts ──────────────────────────────────────────────────────────

section("§1. Contacts")

await step("contacts.list (limit 50) + capture first id", async () => {
  const r: any = await client.contacts.list({ limit: 50 } as any)
  state.contactId = r?.data?.[0]?._id ?? r?.data?.[0]?.id ?? null
  return { count: r?.data?.length ?? 0, sampleId: state.contactId }
})

if (state.contactId) {
  await step("contacts.get(contactId)", () => (client.contacts as any).get(state.contactId!))
  skipped("contacts.update", "would mutate prod contact — destructive")
  await step("contacts.getComments(contactId)", () => (client.contacts as any).getComments(state.contactId!))
  await step("contacts.getFiles(contactId)", () => (client.contacts as any).getFiles(state.contactId!))
} else {
  skipped("contacts.get / update / getComments / getFiles", "no contact fixture in this org")
}

// getActivities takes a conversation_id, not a contact_id. Resolve it from §2.

// ── §2. Conversations ───────────────────────────────────────────────────

section("§2. Conversations")

await step("platform.listBusinessUnits + capture first bu_id", async () => {
  const bus: any = await client.platform.listBusinessUnits()
  state.buId = bus?.[0]?._id ?? bus?.[0]?.id ?? null
  return { count: (bus ?? []).length, sampleId: state.buId }
})

if (state.buId) {
  await step("conversations.search ({ businessUnitId, q='support' })", async () => {
    const r: any = await (client.conversations as any).search({ businessUnitId: state.buId!, q: "support", limit: 20 })
    state.convId = r?.data?.[0]?._id ?? r?.data?.[0]?.id ?? null
    return { count: r?.data?.length ?? 0, sampleConvId: state.convId }
  })
  await step("conversations.getOutstanding ({ businessUnitId, limit: 50 })",
    () => (client.conversations as any).getOutstanding({ businessUnitId: state.buId!, limit: 50 }))
} else {
  skipped("conversations.search / getOutstanding", "no businessUnit fixture")
}

skipped("conversations.assignTeamMember", "would re-assign a real conversation — destructive")
skipped("conversations.updateStatus", "would close a real conversation — destructive")

// Now §1 follow-up: contacts.getActivities(conversationId)
if (state.convId) {
  await step("contacts.getActivities(conversationId)",
    () => (client.contacts as any).getActivities(state.convId!))
} else {
  skipped("contacts.getActivities(conversationId)", "no conversation fixture")
}

// ── §3. Messaging ─────────────────────────────────────────────────────────

section("§3. Messaging")

await step("channel.list", () => client.channel.list())
skipped("messages.send", "would send a real message to a real channel — destructive")

if (state.convId) {
  await step("messages.list (limit 20) — needs conversation_id query param",
    () => (client.messages as any).list?.({ limit: 20, conversation_id: state.convId } as any) ?? Promise.resolve([]))
} else {
  skipped("messages.list", "needs conversation_id fixture")
}

// ── §4. Campaigns & touchpoints ──────────────────────────────────────────

section("§4. Campaigns + touchpoints")

await step("campaign.list", () => (client.campaign as any).list({}))

await step("campaign.create (throwaway)", async () => {
  const c: any = await (client.campaign as any).create({ name: `TestCampaign ${ts}`, type: "email" })
  state.campaignId = c?._id ?? c?.id ?? null
  return { id: state.campaignId }
})

if (state.campaignId) {
  await step("campaign.get(campaignId)", () => (client.campaign as any).get(state.campaignId!))
}

await step("campaign.listTouchpoints", () => (client.campaign as any).listTouchpoints())

if (state.campaignId) {
  await step("campaign.createTouchpoint", async () => {
    const tp: any = await (client.campaign as any).createTouchpoint({
      campaign_id: state.campaignId!, type: "email", delay_days: 3,
    })
    state.touchpointId = tp?._id ?? tp?.id ?? null
    return { id: state.touchpointId }
  })

  if (state.touchpointId) {
    await step("campaign.getTouchpoint(touchpointId)",
      () => (client.campaign as any).getTouchpoint(state.touchpointId!))
    await step("campaign.updateTouchpoint(touchpointId, { delay_days: 5 })",
      () => (client.campaign as any).updateTouchpoint(state.touchpointId!, { delay_days: 5 }))
  }
}

await step("campaign.validateTouchpoint", () =>
  (client.campaign as any).validateTouchpoint({ type: "email", template_id: "tpl_xxx" }),
)

// Touchpoint cleanup
if (state.touchpointId) {
  await step("campaign.deleteTouchpoint (cleanup)",
    () => (client.campaign as any).deleteTouchpoint(state.touchpointId!))
}
// Campaign cleanup
if (state.campaignId) {
  await step("campaign.delete (cleanup)",
    () => (client.campaign as any).delete(state.campaignId!))
}

// ── §5. Message suggestion ───────────────────────────────────────────────

section("§5. Message suggestion")
if (state.convId) {
  await step("messageSuggestion.getSuggestions ({ conversation_id, limit: 3 })", () =>
    (client.messageSuggestion as any).getSuggestions({ conversation_id: state.convId!, limit: 3 }),
  )
} else {
  skipped("messageSuggestion.getSuggestions", "no conversation fixture")
}

// ── §6. Predict ──────────────────────────────────────────────────────────

section("§6. Predict")
note("backend-known-issue: predict.predict is org-config dependent — model 'lead_score_v1' may not be deployed for this org")
await step("predict.predict ({ model: 'lead_score_v1', input: { ... } })", () =>
  (client.predict as any).predict({
    model: "lead_score_v1",
    input: { company_size: 200, industry: "saas", mrr: 5000 },
  }),
)

// ── Summary ──────────────────────────────────────────────────────────────

console.log(`\n━━━ Summary (resources / API key) ━━━`)
console.log(`pass=${pass}  fail=${fail}  skip=${skip}`)
if (fails.length) { console.log("Failures:"); fails.forEach(f => console.log(`  - ${f}`)) }
if (docGaps.length) { console.log("Doc / backend gaps:"); docGaps.forEach(g => console.log(`  - ${g}`)) }
process.exit(fail > 0 ? 1 : 0)

File: test-api-pkg/py/test_docs_resources.py
"""Mirrors website/public/sdk/resources.md against `imbrace==1.0.4` (PyPI)
— API-key auth.

Many resources here (contacts/channel-service/campaign/predict) are flagged
as unreachable on app-gatewayv2 in FIX_PLAN_v1.0.6.md §A.1 — most failures
below are backend ops issues, not SDK bugs.
"""
from __future__ import annotations
import json
import os
import sys
import time

from dotenv import load_dotenv
load_dotenv()

from imbrace import ImbraceClient

API_KEY = os.environ.get("IMBRACE_API_KEY")
ORG_ID  = os.environ.get("IMBRACE_ORGANIZATION_ID")
GATEWAY = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

if not API_KEY or not ORG_ID:
    print("Missing IMBRACE_API_KEY or IMBRACE_ORGANIZATION_ID")
    sys.exit(1)

client = ImbraceClient(api_key=API_KEY, organization_id=ORG_ID, gateway=GATEWAY, timeout=30)

passed = failed = skipped_n = 0
failures: list[str] = []
doc_gaps: list[str] = []


def step(label, fn, expect_fail=False):
    global passed, failed
    sys.stdout.write(f"  - {label} ... "); sys.stdout.flush()
    try:
        t0 = time.time()
        r = fn()
        dt = int((time.time() - t0) * 1000)
        summary = json.dumps(r, default=str)[:90] if r is not None else ""
        if expect_fail:
            print(f"unexpected pass [{dt}ms] {summary}")
            failed += 1; failures.append(f"{label} -> unexpected pass")
        else:
            print(f"OK [{dt}ms] {summary}"); passed += 1
    except Exception as e:
        code = str(e)[:80]
        if expect_fail:
            print(f"OK (expected fail [{code}])"); passed += 1
        else:
            print(f"FAIL [{code}]"); failed += 1
            failures.append(f"{label} -> {code}")


def skip(label, reason):
    global skipped_n
    print(f"  - {label}  SKIP: {reason}"); skipped_n += 1


def section(title): print(f"\n== {title} ==")
def note(msg): print(f"  i {msg}"); doc_gaps.append(msg)


print(f"\n=== DOCS: resources.md - auth: API KEY (PyPI imbrace==1.0.4) ===")
print(f"gateway={GATEWAY}\norg={ORG_ID}\n")

ts = int(time.time() * 1000)
state = {"contactId": None, "convId": None, "buId": None, "campaignId": None, "touchpointId": None}

note("backend-known-issue: contacts/channel-service/campaign/predict often timeout on app-gatewayv2 (FIX_PLAN_v1.0.6 §A.1)")


section("§1. Contacts")


def _list_contacts():
    r = client.contacts.list(limit=50)
    data = r.get("data", []) if isinstance(r, dict) else (r or [])
    state["contactId"] = data[0].get("_id") if data else None
    return {"count": len(data), "sampleId": state["contactId"]}


step("contacts.list (limit=50) + capture first id", _list_contacts)

if state["contactId"]:
    step("contacts.get(contact_id)", lambda: client.contacts.get(state["contactId"]))
    skip("contacts.update", "would mutate prod contact - destructive")
    step("contacts.get_comments(contact_id)", lambda: client.contacts.get_comments(state["contactId"]))
    step("contacts.get_files(contact_id)", lambda: client.contacts.get_files(state["contactId"]))
else:
    skip("contacts.get / update / get_comments / get_files", "no contact fixture")


section("§2. Conversations")


def _list_bu():
    bus = client.platform.list_business_units()
    state["buId"] = (bus[0] or {}).get("_id") if bus else None
    return {"count": len(bus or []), "sampleId": state["buId"]}


step("platform.list_business_units + capture first bu_id", _list_bu)

if state["buId"]:
    def _search_conv():
        r = client.conversations.search(business_unit_id=state["buId"], q="support", limit=20)
        data = r.get("data", []) if isinstance(r, dict) else []
        state["convId"] = data[0].get("_id") if data else None
        return {"count": len(data), "sampleConvId": state["convId"]}
    step("conversations.search (business_unit_id=, q='support')", _search_conv)
else:
    skip("conversations.search", "no businessUnit fixture")

step("conversations.list (limit 50) — Py docs use list (TS uses getOutstanding)",
     lambda: client.conversations.list(limit=50))

skip("conversations.update_status / assign_team_member", "destructive")

if state["convId"]:
    step("contacts.get_activities(conversation_id)",
         lambda: client.contacts.get_activities(state["convId"]))
else:
    skip("contacts.get_activities(conversation_id)", "no conversation fixture")


section("§3. Messaging")
step("channel.list", lambda: client.channel.list())
skip("messages.send", "destructive")
if state["convId"]:
    if hasattr(client.messages, "list"):
        step("messages.list (limit 20, conversation_id=)",
             lambda: client.messages.list(limit=20, conversation_id=state["convId"]))
    else:
        skip("messages.list", "method not exposed on Py SDK 1.0.4")
else:
    skip("messages.list", "needs conversation_id fixture")


section("§4. Campaigns + touchpoints")
step("campaign.list", lambda: client.campaign.list())


def _create_campaign():
    c = client.campaign.create({"name": f"TestCampaign {ts}", "type": "email"})
    state["campaignId"] = c.get("_id") if isinstance(c, dict) else getattr(c, "_id", None)
    return {"id": state["campaignId"]}


step("campaign.create (throwaway)", _create_campaign)

if state["campaignId"]:
    step("campaign.get(campaign_id)", lambda: client.campaign.get(state["campaignId"]))

step("campaign.list_touchpoints", lambda: client.campaign.list_touchpoints())

if state["campaignId"]:
    def _create_touchpoint():
        tp = client.campaign.create_touchpoint({
            "campaign_id": state["campaignId"], "type": "email", "delay_days": 3,
        })
        state["touchpointId"] = tp.get("_id") if isinstance(tp, dict) else getattr(tp, "_id", None)
        return {"id": state["touchpointId"]}
    step("campaign.create_touchpoint", _create_touchpoint)

    if state["touchpointId"]:
        step("campaign.get_touchpoint(touchpoint_id)",
             lambda: client.campaign.get_touchpoint(state["touchpointId"]))
        step("campaign.update_touchpoint(touchpoint_id, { delay_days: 5 })",
             lambda: client.campaign.update_touchpoint(state["touchpointId"], {"delay_days": 5}))

step("campaign.validate_touchpoint",
     lambda: client.campaign.validate_touchpoint({"type": "email", "template_id": "tpl_xxx"}))

if state["touchpointId"]:
    step("campaign.delete_touchpoint (cleanup)",
         lambda: client.campaign.delete_touchpoint(state["touchpointId"]))
if state["campaignId"]:
    step("campaign.delete (cleanup)", lambda: client.campaign.delete(state["campaignId"]))


section("§5. Message suggestion")
if state["convId"]:
    step("message_suggestion.get_suggestions ({ conversation_id, limit: 3 })",
         lambda: client.message_suggestion.get_suggestions({"conversation_id": state["convId"], "limit": 3}))
else:
    skip("message_suggestion.get_suggestions", "no conversation fixture")


section("§6. Predict")
note("backend-known-issue: predict.predict org-config dependent — model 'lead_score_v1' may not be deployed for this org")
step("predict.predict ({ model: 'lead_score_v1', input })",
     lambda: client.predict.predict({
         "model": "lead_score_v1",
         "input": {"company_size": 200, "industry": "saas", "mrr": 5000},
     }))


print(f"\n=== Summary (resources / API key) ===")
print(f"pass={passed}  fail={failed}  skip={skipped_n}")
if failures:
    print("Failures:")
    for f in failures: print(f"  - {f}")
if doc_gaps:
    print("Doc / backend gaps:")
    for g in doc_gaps: print(f"  - {g}")
sys.exit(1 if failed > 0 else 0)
